compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -2799,6 +2805,27 @@ private void markAndRegisterClosureVariable(BSymbol symbol, DiagnosticPos pos) {
                 ((BLangArrowFunction) env.node).closureVarSymbols.add(new ClosureVarSymbol(resolvedSymbol, pos));
             }
         }
+
+        BLangNode node = env.node;
+        SymbolEnv cEnv = env;
+        while (node != null && node.getKind() != NodeKind.FUNCTION) {
+            if (node.getKind() == NodeKind.TRANSACTION) {
+                SymbolEnv encInvokableEnv = findEnclosingInvokableEnv(env, encInvokable);
+                BSymbol resolvedSymbol = symResolver.lookupClosureVarSymbol(encInvokableEnv, symbol.name,
+                        SymTag.VARIABLE);
+                if (resolvedSymbol != symTable.notFoundSymbol) {
+                    resolvedSymbol.closure = true;
+                }
+                break;
+            } else {
+                SymbolEnv enclEnv = cEnv.enclEnv;
+                if (enclEnv == null) {
+                    break;
+                }
+                cEnv = enclEnv;
+                node = cEnv.node;
+            }
```

Comment:
Why we need else part here? Also shall we add comments for whole logic.

Reply:
@warunalakshitha 
What the logic does is, check if the variable is used inside a transaction function to mark it as closure instead of marking all the variables inside transaction as a closure. 

The `else` part is required to traverse through parent node until a function node is met.

I have added a comment above the logic.